tests/integration/run_modes/test_mode_overrides.py

```diff
@@ -0,0 +1,152 @@
+# THIS FILE IS PART OF THE CYLC WORKFLOW ENGINE.
+# Copyright (C) NIWA & British Crown (Met Office) & Contributors.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Test that using [runtime][TASK]run mode works in each mode.
+
+Point 3 of the Skip Mode proposal
+https://github.com/cylc/cylc-admin/blob/master/docs/proposal-skip-mode.md
+
+| The run mode should be controlled by a new task configuration
+| [runtime][<namespace>]run mode with the default being live.
+| As a runtime configuration, this can be defined in the workflow
+| for development / testing purposes or set by cylc broadcast.
+
+n.b: This is pretty much a functional test and
+probably ought to be labelled as such, but uses the
+integration test framework.
+"""
+
+import pytest
+
+
+@pytest.mark.parametrize(
+    'workflow_run_mode', [('live'), ('skip')])
+async def test_run_mode_override_from_config(
+    workflow_run_mode, flow, scheduler, run, complete, log_filter
+):
+    """Test that ``[runtime][TASK]run mode`` overrides workflow modes.
+    """
+    cfg = {
+        "scheduler": {"cycle point format": "%Y"},
+        "scheduling": {
+            "initial cycle point": "1000",
+            "final cycle point": "1000",
+            "graph": {"P1Y": "live_\nskip_\ndefault_"}},
+        "runtime": {
+            "skip_": {"run mode": "skip"},
+            "live_": {"run mode": "live"}
+        }
+    }
+    id_ = flow(cfg)
+    schd = scheduler(id_, run_mode=workflow_run_mode, paused_start=False)
```

Comment:
This integration test is actually running real jobs!

However, I don't think it's actually testing that the `run mode` is correctly overridden because it doesn't check how the task is submitted.

Also, it's testing `live` and `skip` modes, however, the workflow run modes are `live` and `simulation`?!

Here's my suggestion, it goes with the small `task_job_mgr` refactor I posted above (makes it easier to suppress live jobs from running whilst still testing non-live submission):

```python
from cylc.flow.task_state import TASK_STATUS_SUBMITTED, TASK_STATUS_SUCCEEDED

@pytest.fixture
def capture_live_submissions(capcall):
    """Capture live submission attempts.

    This prevents real jobs from being submitted to the system.

    If you call this fixture from a test, it will return a set of tasks that
    would have been submitted had this fixture not been used.
    """
    def fake_submit(self, _workflow, itasks, *_):
        for itask in itasks:
            for status in (TASK_STATUS_SUBMITTED, TASK_STATUS_SUCCEEDED):
                self.task_events_mgr.process_message(
                    itask,
                    'INFO',
                    status,
                    '2000-01-01T00:00:00Z',
                    '(received)',
                )
        return itasks

    # suppress and capture live submissions
    submit_live_calls = capcall(
        'cylc.flow.task_job_mgr.TaskJobManager._submit_live_task_jobs',
        fake_submit
    )

    def get_submissions():
        nonlocal submit_live_calls
        return {
            itask.identity
            for ((_self, _workflow, itasks, *_), _kwargs) in submit_live_calls
            for itask in itasks
        }

    return get_submissions


@pytest.mark.parametrize('workflow_run_mode', ['live', 'skip'])  # TODO: Should this be 'simulation'?
async def test_run_mode_override_from_config(
    capture_live_submissions, flow, scheduler, run, complete, workflow_run_mode
):
    """Test that `[runtime][<namespace>]run mode` overrides workflow modes."""
    id_ = flow({
        'scheduling': {
            'graph': {
                'R1': 'live & skip',
            },
        },
        'runtime': {
            'live': {'run mode': 'live'},
            'skip': {'run mode': 'skip'},
        }
    })
    schd = scheduler(id_, run_mode=workflow_run_mode, paused_start=False)
    async with run(schd):
        await complete(schd)

    if workflow_run_mode == 'live':
        assert capture_live_submissions() == {'1/live'}
    else:
        assert capture_live_submissions() == set()
```

Note: This test requires https://github.com/cylc/cylc-flow/pull/6326

Could go further by mocking away the `get_submission_method` method to test which of the dummy submission methods is called if desired.

For bonus robustness, we could also do this:

```python
@pytest.mark.parametrize('workflow_run_mode', [mode.value for mode in WORKFLOW_MODES])
```

To prevent the test from getting out of sync with changes to workflow run modes.

Reply:
> Also, it's testing `live` and `skip` modes, however, the workflow run modes are `live` and `simulation`?!

These are the task modes which you can override by broadcasting.

Reply:
There were a couple of interesting problems with this (which I have solved):

1. There was not previously a mockable method called `_submit_live_task_jobs`. I have now created something like it in the form of `_submit_livelike_task_jobs`.
2. `WORKFLOW_MODES` is a frozenset, which is not ordered. Pytest-xdist doesn't like parameters without a fixed order, so a `sorted(WORKFLOW_MODES)` was required. [Source](https://pytest-xdist.readthedocs.io/en/latest/known-limitations.html#order-and-amount-of-test-must-be-consistent).